src/wolframite/runtime/system.clj

```diff
@@ -0,0 +1,161 @@
+(ns wolframite.runtime.system
+  "For dealing with the runtime system, primarily the operating system specific defaults."
+  (:require
+   [babashka.fs :as fs]
+   [clojure.string :as string])
+  (:import
+   (java.lang System)))
+
+(def supported-OS #{:linux :mac :windows})
+
+;; Default path information can be listed here, in order of preference.
+(def ^:private defaults
+  [{:root  "/usr/local/Wolfram/Mathematica"
+    :kernel "Executables/MathKernel"
+    :product :mathematica
+    :os :linux}
+   {:root  "/opt/Mathematica"
+    :kernel "Executables/MathKernel"
+    :product :mathematica
+    :os :linux}
+   {:root  "/usr/local/Wolfram/WolframEngine"
+    :kernel "Executables/WolframKernel"
+    :product :wolfram-engine
+    :os :linux}
+
+   {:root  "/Applications/Mathematica.app/Contents"
+    :kernel "MacOS/MathKernel"
+    :product :mathematica
+    :os :mac}
+   {:root   "/Applications/Wolfram Engine.app/Contents/Resources/Wolfram Player.app/Contents"
+    :kernel "MacOS/WolframKernel"
+    :product :wolfram-engine
+    :os :mac}
+
+   {:root "/c:/Program Files/Wolfram Research/Mathematica/"
+    :kernel "WolframKernel.exe"
+    :product :mathematica
+    :os :windows}
+   {:root "/c:/Program Files/Wolfram Research/Wolfram Engine/"
+    :kernel "MathKernel.exe"
+    :product :wolfram-engine
+    :os :windows}])
+
+(defn- user-paths
+  "The paths given via Java property or environment variables.
+
+  NOTE: These should be root directories that distinguish Mathematica or Wolfram engine files, not subdirectories. E.g.  '/usr/local/Wolfram/Mathematica'.
+
+  TODO:
+  - (jh) alternatively, support Java system properties?"
+  []
+  (let [names  ["JLINK_JAR_PATH"
+                "MATHEMATICA_INSTALL_PATH"
+                "WOLFRAM_INSTALL_PATH"]
+        method (fn [f] (-> f (nth  2) first str (string/split #"/") last))
+        fs ['(fn [name] (System/getProperty name)) '(fn [name] (System/getenv name))]]
+    (->> (map (fn [f]
+                (map (fn [name] [[(keyword name) ((eval f) name)]
+                                 [:source (method f)]])
+                     names))
+              fs)
+         (apply concat)
+         (map #(into {} %)))))
+
+(defn- version-number?
+  [path]
+  (-> (fs/file-name path)
+      first
+      str
+      parse-long
+      some?))
+
+(defn- version
+  "Extracts the version from a given path into a vector of numbers.
+
+  ASSUME: path node (folder) is period-separated number, e.g. 13.0.2 -> [13 0 2]
+  "
+  [version-path]
+  (-> version-path
+      fs/file-name
+      (string/split #"\.")
```

Comment:
traditionally, you require clojure.string as str.

Also, what will happen if the file name is not .-separated numbers? Is blowing up good enough? Or should we be more careful and check with e.g. `#(re-matches #"(\d+)(?:\.(\d+))*" %)`?

Reply:
Regex is better still, but this was an oversight and done better in the latest patch.

Reply:
On the 'str' vs 'string', I'll change it for consistency, but in general I think this is a mistake. I take the code is for reading rather than for writing (and if you're not using autocompletion then you're wrong :) ) approach.